=str #16986 Handle early cancelation issue

patriknw opened the pull request:
and =str #15191 Verify that stage actors are stopped

Some more things to verify and add, but you can take a first look and see if it is on right track. Note 2 commits if you want to review the two tickets in isolation.

patriknw committed:
=str #15191 Verify that stage actors are stopped

* found one bug of too early actor stop and thereby missing cancel of upstream,
  in fan-in tests ""work with one immediately failed and one nonempty publisher"

patriknw committed:
=str #16986 Handle early cancelation issue

* if downstream cancels before subscription from upstream is
  received the cancel was not propagated to upstream, the
  actor was stopped too early

patriknw commented:
I think this PR is complete now. Please review.

Used checkThatAllStagesAreStopped in many tests.

Found one more bug of too early actor stop and thereby missing cancel of upstream , in fan-in tests ""work with one immediately failed and one nonempty publisher".

ktoso commented:
LGTM

rkuhn commented on akka-stream-testkit/src/test/scala/akka/stream/testkit/StreamTestKit.scala:
```diff
@@ -174,4 +178,20 @@ object StreamTestKit {
+        val probe = TestProbe()(impl.system)
+        probe.awaitAssert {
+          impl.supervisor.tell(StreamSupervisor.GetChildren, probe.ref)
+          val children = probe.expectMsgType[StreamSupervisor.Children].children
```
might want to set a small timeout on this per-iteration expectation

rkuhn replied:
or not

patriknw replied:
not, since I know that StreamSupervisor will reply immediately

rkuhn replied:
yup, I figured ;-)

rkuhn commented:
LGTM! Will be fun to rebase TLS on top of that :-)

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala:
```diff
@@ -222,6 +224,13 @@ private[akka] object StreamSupervisor {
   final case class Materialize(props: Props, name: String) extends DeadLetterSuppression
+
+  /** Testing purpose */
+  final case object GetChildren
```
So the thing is, I believe we can make this a proper API. We need to discuss the details, I have some ideas. Good enough for this PR though, just wanted to mention.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/Transfer.scala:
```diff
@@ -146,9 +154,31 @@ private[akka] trait Pump {
-  final def nextPhase(phase: TransferPhase): Unit = {
-    transferState = phase.precondition
-    currentAction = phase.action
+  final def initialPhase(waitForUpstream: Int, andThen: TransferPhase): Unit = {
```
nice one!

drewhk commented:
LGTM

patriknw commented:
rebased, and fixed review comments, such as rename to `assertAllStagesStopped`

patriknw commented:
real test failure that must be investigated before merging

rkuhn commented:
real shame, I wanted to use this in the TLS tests ;-)

patriknw commented:
I found the bug. You can review it in isolation in 8012cd4.
I will rebase.

patriknw commented:
rebased again, and added one more test as requested by @drewhk : https://github.com/akka/akka/pull/17217/files#diff-59a559d890e855094590029696178df5R397

patriknw commented:
We might have a problem with the TLS stage also as I mentioned to @rkuhn
I will not have time to try that today, so if this becomes green I will merge it.

patriknw commented:
wat, did the target branch change again :(

drewhk commented:
Lgtm
2015.04.23. 19:55 ezt írta ("Patrik Nordwall" <notifications@github.com>):

> rebased again, and added one more test as requested by @drewhk
> <https://github.com/drewhk> :
> https://github.com/akka/akka/pull/17217/files#diff-59a559d890e855094590029696178df5R397
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/17217#issuecomment-95670146>.
>

patriknw commented:
the failures was in ConnectionPoolSpec, reported elsewhere

patriknw commented:
Refs #16986 and #15191
